### PR TITLE
add microcoap example application

### PR DESCRIPTION
**Pull request by Lotterleben**
Add a small microcoap example server along with instructions on how to test it using [marz](https://github.com/sgso/marz) and the Copper Firefox Plugin. Depends on https://github.com/RIOT-OS/RIOT/pull/2383.

### CODE REVIEW

**Comment by gebart** [comment]
I can't build this example:

```
applications/microcoap/endpoints.c:10:13: error: variably modified ‘response’ at file scope
 static char response[response_len] = "";
             ^
```

arm-none-eabi-gcc (Gentoo 4.8.3 p1.1, pie-0.5.9) 4.8.3

When I change it to numeric constant `1500` I get the program to build correctly.

**Comment by Lotterleben** [comment]
Yup, that has been fixed, but on the wrong branch it seems. 1 sec.

**Comment by Lotterleben** [comment]
Should build now :)

File: microcoap/Makefile [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+# If no BOARD is found in the environment, use this default:
+BOARD ?= native
+
+# This has to be the absolute path to the RIOT base directory:
+RIOTBASE ?= $(CURDIR)/../../RIOT
+
+# Comment this out to disable code in RIOT that does safety checking
+# which is not needed in a production environment but helps in the
+# development process:
+CFLAGS += -DRIOT -DMICROCOAP_DEBUG #-DENABLE_DEBUG
```
Comment (authmillenon) [inline microcoap/Makefile]: Remove the out-commented flag

File: microcoap/Makefile [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+USEMODULE += config
+USEMODULE += shell
+USEMODULE += shell_commands
+USEMODULE += uart0
+
+ifeq ($(strip $(BOARD)),native)
+	USEMODULE += nativenet
+else
+	USEMODULE += cc110x
+endif
```
Comment (authmillenon) [inline microcoap/Makefile]: What about boards, that support neither?

File: microcoap/main.c [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+static void _init_tlayer(void)
+{
+    msg_init_queue(msg_q, RCV_MSG_Q_SIZE);
+
+    net_if_set_hardware_address(0, get_hw_addr());
+
+    printf("initializing 6LoWPAN...\n");
+
+    ipv6_addr_init(&prefix, 0xABCD, 0xEF12, 0, 0, 0, 0, 0, 0);
+    if_id = 0; // >1 interface isn't supported anyway, so there
```
Comment (authmillenon) [inline microcoap/main.c]: Use C-style comments

File: microcoap/main.c [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    _init_tlayer();
+    thread_create(_rcv_stack_buf, KERNEL_CONF_STACKSIZE_MAIN, PRIORITY_MAIN, CREATE_STACKTEST, _microcoap_server_thread, NULL ,"_microcoap_server_thread");
+
+    /* Open the UART0 for the shell */
+    posix_open(uart0_handler_pid, 0);
+
+    printf("\n\t\t\tWelcome to RIOT\n\n");
+
+    shell_t shell;
+    shell_init(&shell, NULL, UART0_BUFSIZE, uart0_readc, uart0_putc);
```
Comment (authmillenon) [inline microcoap/main.c]: Not sure why you start a shell for a server.

Comment (Lotterleben) [inline microcoap/main.c]: Aaaah that's somewhat legacy– you have to tell marz the hwaddr by setting in the marz.config– and at first I didn't want to set it to 1 because that sounded like a recipe for desaster to me so I added shell functionality which lets you query for the (randomly assigned) hwaddr so you can add that to your config. That turned out to be pretty annoying so I abandoned it. I actually don't like either of the approaches, but you're right, I'll delete the shell code.

Comment (authmillenon) [inline microcoap/main.c]: What I usually do is to use a hash of the CPU ID of the MTU as hardware address ;) For native this is more or less random at every restart (except you set the ID with `-i`) because the CPU ID is based on the PID of the native process.

Comment (LudwigOrtmann) [inline microcoap/main.c]: Why don't you just specify the address at build time?

Comment (Lotterleben) [inline microcoap/main.c]: @authmillenon @LudwigOrtmann Don't both approaches boild down to the same problems, too– either I'll have a random address that i'll always have to adjust in the marz config, or I'll have one fixed address, which will cause problems as soon as I start more than one instance of this app? (which wouldn't make sense, though)
Or am I missing something here?

File: microcoap/Makefile (hunk 2 above)
Comment (Lotterleben) [inline microcoap/Makefile]: Actually, when I think about it, does it even make sense to include anything but native for the example? Having just native makes for a much slimmer Makefile, and I can't guarantee (yet) that it will work on any proper boards anyway. If you think this application should reflect our experiences (i.e. we definitely got this to work on the SAMR21, or something like that), I'd add the lines for that as soon as I'm confident about them.  
TL;DR: I propose to get rid of anything but native for now.

File: microcoap/main.c (hunk 4 above)
Comment (authmillenon) [inline microcoap/main.c]: With the CPU ID it is randomized, but for each instance it will stay the same: say you have two samr21-xpro boards, than for board 1 the address will always be e.g. `0x4a2c` and for board 2 it will always be `0x5fb4`. Only for native where you effectively "destroy" your board everytime you kill the process it is running in the ID changes. But you can circumvent this with the `-i` parameter to your TERMFLAGS. E. g.:

```
$ ./examples/default/bin/native/default.elf tap0
[...]
CPU ID: 000000 82 33 00 00
$ ./examples/default/bin/native/default.elf tap0
[...]
CPU ID: 000000 86 33 00 00
$ ./examples/default/bin/native/default.elf tap0 -i 1234
[...]
CPU ID: 000000 d2 04 00 00
$ ./examples/default/bin/native/default.elf tap0 -i 1234
[...]
CPU ID: 000000 d2 04 00 00
```

Problem is: for a lot of the older boards (e.g. the msba2) this feature is not implemented yet.

Comment (authmillenon) [inline microcoap/main.c]: (Had to edit the output a little bit, since I copy-pasted it wrong together ;-))

File: microcoap/Makefile (hunk 2 above)
Comment (authmillenon) [inline microcoap/Makefile]: :+1:

Comment (authmillenon) [inline microcoap/Makefile]: (use `WHITELIST_BOARDS` for that) ;-)

File: microcoap/main.c (hunk 4 above)
Comment (Lotterleben) [inline microcoap/main.c]: ...But what's the benefit, then?

Comment (LudwigOrtmann) [inline microcoap/main.c]: For native/marz, if you need to know the address in advance, you can, without adding a shell, by specifying the `ID`.

For the sam, do you need to know the hardware address?

Comment (Lotterleben) [inline microcoap/main.c]: Oooh okay, so I'd have a flexible hwaddr where I need it (when this app runs on an actual board) and a (somewhat) predictable one when I need it, too (on native).

So, suggested improvements:  
- set hwaddr to a hash of the CPU ID of the MTU as hardware address
- remove shell code
- add instruction to specify interface as set in marz.config using the -i parameter

okay?

Comment (LudwigOrtmann) [inline microcoap/main.c]: Mkay.

**Comment by Lotterleben** [comment]
@LudwigOrtmann @authmillenon 
I'm sorry, but I'm gong to have to ask *again*– I re-implemented ``get_hw_addr()`` now as suggested, but  I'm still not content with the way it is now– the user may not have to constantly adjust the hwaddr in marz.config if they use -i, but they'll have to do it once (because I have no idea for which -i ``get_hw_addr()`` will return 1, which is the default hwaddr i marz.config).
And I just don't get what's wrong with just returning sysconfig.id and telling the user to start the native instance which they want to be connected to marz with``-i 1``?

**Comment by Lotterleben** [comment]
Aaaah, because it will break other things -.- *sighs*

**Comment by authmillenon** [comment]
Looks good, please squash.

**Comment by Lotterleben** [comment]
Done :)

**Comment by authmillenon** [comment]
```
"make" -C /home/martine/Repositories/RIOT-OS/RIOT/pkg/microcoap/microcoap
/home/martine/Repositories/RIOT-OS/applications/microcoap/endpoints.c:10:13: error: variably modified ‘response’ at file scope
 static char response[response_len] = "";
```

with gcc

```
"make" -C /home/martine/Repositories/RIOT-OS/RIOT/pkg/microcoap/microcoap
/home/martine/Repositories/RIOT-OS/applications/microcoap/endpoints.c:10:13: error: variable length array declaration not
      allowed at file scope
static char response[response_len] = "";
            ^        ~~~~~~~~~~~~
/home/martine/Repositories/RIOT-OS/applications/microcoap/endpoints.c:20:80: warning: unused parameter 'inpkt'
      [-Wunused-parameter]
static int handle_get_response(coap_rw_buffer_t *scratch, const coap_packet_t *inpkt, coap_packet_t *outpkt, uint8_t id...
                                                                               ^
/home/martine/Repositories/RIOT-OS/applications/microcoap/endpoints.c:20:118: warning: unused parameter 'id_hi'
      [-Wunused-parameter]
  ...handle_get_response(coap_rw_buffer_t *scratch, const coap_packet_t *inpkt, coap_packet_t *outpkt, uint8_t id_hi, uint8_...
                                                                                                               ^
/home/martine/Repositories/RIOT-OS/applications/microcoap/endpoints.c:20:133: warning: unused parameter 'id_lo'
      [-Wunused-parameter]
  ...*scratch, const coap_packet_t *inpkt, coap_packet_t *outpkt, uint8_t id_hi, uint8_t id_lo)
                                                                                         ^
3 warnings and 1 error generated.
```

with clang

**Comment by authmillenon** [comment]
```diff
diff --git a/microcoap/endpoints.c b/microcoap/endpoints.c
index 0c8869e..f6b6967 100644
--- a/microcoap/endpoints.c
+++ b/microcoap/endpoints.c
@@ -6,14 +6,14 @@
 #define ENABLE_DEBUG    (0)
 #include "debug.h"
 
-uint16_t response_len = 1500;
-static char response[response_len] = "";
+#define RESPONSE_LEN   (1500)
+static char response[RESPONSE_LEN] = "";
 
 static const coap_endpoint_path_t path = {2, {"foo", "bar"}};
 
 void create_response(void)
 {
-    strncat(response, "1337", response_len-5);
+    strncat(response, "1337", RESPONSE_LEN-5);
 }
 
 /* The handler which handles the path /foo/bar */
```

fixes it.

**Comment by Lotterleben** [comment]
That one seems to haunt me :D This was already fixed, but I must've lost it in the rebase... Sorry. 1sec.

**Comment by Lotterleben** [comment]
Wait.. How did you get that? I can't seem to find the offending line(s) in the current code.

**Comment by authmillenon** [comment]
Arghs checked out wrong branch :blush:

**Comment by authmillenon** [comment]
All is fine: ACK

**Comment by Lotterleben** [comment]
:sparkles: Thanks for reviewing! :)
